metals/src/main/scala/scala/meta/internal/metals/testProvider/frameworks/JunitTestFinder.scala

```diff
@@ -3,46 +3,52 @@ package scala.meta.internal.metals.testProvider.frameworks
 import scala.meta.internal.metals.MetalsEnrichments._
 import scala.meta.internal.metals.testProvider.TestCaseEntry
 import scala.meta.internal.mtags
+import scala.meta.internal.mtags.Semanticdbs
 import scala.meta.internal.semanticdb.SymbolInformation
 import scala.meta.internal.semanticdb.TextDocument
 import scala.meta.internal.semanticdb.TypeRef
 import scala.meta.io.AbsolutePath
 
-class JunitTestFinder {
+class JunitTestFinder(semanticdbs: Semanticdbs) {
   import JunitTestFinder._
 
   def findTests(
-      doc: TextDocument,
+      doc: Option[TextDocument],
       path: AbsolutePath,
       suiteSymbol: mtags.Symbol
-  ): List[TestCaseEntry] = {
-    val uri = path.toURI
-
-    def isMethodWithTestAnnotation(symbol: SymbolInformation) =
-      symbol.kind == SymbolInformation.Kind.METHOD && symbol.annotations
-        .exists(_.tpe match {
-          case TypeRef(_, annotationSymbol, _) =>
-            annotationSymbol == junitAnnotationSymbol
-          case _ => false
-        })
-
-    def isValid(symbol: SymbolInformation): Boolean = {
-      isMethodWithTestAnnotation(symbol) && symbol.symbol.startsWith(
-        suiteSymbol.value
-      )
-    }
-
-    doc.symbols
-      .collect {
-        case symbol if isValid(symbol) =>
-          doc
-            .toLocation(uri.toString, symbol.symbol)
-            .map(location => TestCaseEntry(symbol.displayName, location))
+  ): Vector[TestCaseEntry] = {
+    // if text document isn't defined try to fetch it from semanticdbs
```

Comment:
Yes. JUnit tests (all tests in general, but currently only junit needs semanticDB) should be discovered in 3 cases:
1. File was compiled and it's opened - semanticdb will be defined
2. Suite is discovered for the first time and the file is opened - semanticDB has to be fetched because there isn't any.
3. File which contains suites was opened - good UX requires discovering tests and showing them to the user.

While I was writing this I realized that I could use `Trees` for JUnit too. However, maybe I'll leave it as it is for now, who knows what other frameworks will need

Reply:
oh, now I see.

IMO, it's a bit misleading that  `getTestCasesForPath` and `getTestCasesForSuites` in `TestSuitesProvider` takes `Option[TextDocument]` and in these cases when it's called with `None` it leads to this fallback on`semanticdbs`.

> While I was writing this I realized that I could use Trees for JUnit too. However, maybe I'll leave it as it is for now, who knows what other frameworks will need

I think `TextDocument` should help  with a case with tests defined in parent class.
Using it we can obtain a signature of class and take its parents.

Reply:
> IMO, it's a bit misleading that getTestCasesForPath and getTestCasesForSuites in TestSuitesProvider takes Option[TextDocument] and in these cases when it's called with None it leads to this fallback onsemanticdbs.

Do you have a suggestion on how to reduce this misleading?

Reply:
Maybe I can put this default back in the `TestSuitesProvider` and add docstring?

Reply:
Ok. I took look and think that at least for now it's valid and allow to avoid unnecessary semanticdb loading from disk.
Let's keep it as it is.